src/github.com/matrix-org/dendrite/roomserver/storage/sql.go

```diff
@@ -68,3 +103,267 @@ func (s *statements) upsertPartitionOffset(topic string, partition int32, offset
 	_, err := s.upsertPartitionOffsetStmt.Exec(topic, partition, offset)
 	return err
 }
+
+func (s *statements) prepareEventTypes(db *sql.DB) (err error) {
+	_, err = db.Exec(eventTypesSchema)
+	if err != nil {
+		return
+	}
+	if s.insertEventTypeNIDStmt, err = db.Prepare(insertEventTypeNIDSQL); err != nil {
+		return
+	}
+	if s.selectEventTypeNIDStmt, err = db.Prepare(selectEventTypeNIDSQL); err != nil {
+		return
+	}
+	return
+}
+
+const eventTypesSchema = `
+-- Numeric versions of the event "type"s. Event types tend to be taken from a
+-- small common pool. Assigning each a numeric ID should reduce the amount of
+-- data that needs to be stored and fetched from the database.
+-- It also means that many operations can work with int64 arrays rather than
+-- string arrays which may help reduce GC pressure.
```

Comment:
> Assigning each a numeric ID **should** reduce the amount of data that needs to be stored and fetched from the database. It also means that many operations can work with int64 arrays rather than string arrays which **may** help reduce GC pressure.

This feels like a massive premature optimisation. You're going to need to give me some actual evidence to support your theories: benchmarks done by yourself or others, or something better than just a hand-wavey comment.

I dislike this because look at how much of this PR is dedicated to doing this mapping, and no real "useful" work. All the `assign*` functions and these schema files just add boilerplate which I'd rather not have unless there was an extremely compelling reason for it. I don't find these comments compelling enough.